src/argus/notificationprofile/media/base.py

```diff
@@ -3,62 +3,142 @@
 from abc import ABC, abstractmethod
 from typing import TYPE_CHECKING
 
+from django import forms
+from django.core.exceptions import ValidationError
+
+from ..models import DestinationConfig
+
 if TYPE_CHECKING:
     from collections.abc import Iterable
 
     from types import NoneType
-    from typing import Union
+    from typing import Union, Generator
 
     from django.contrib.auth import get_user_model
     from django.db.models.query import QuerySet
 
     from argus.incident.models import Event
-    from ..models import DestinationConfig
-    from ..serializers import RequestDestinationConfigSerializer
 
     User = get_user_model()
 
 
 __all__ = ["NotificationMedium"]
 
 
+class CommonDestinationConfigForm(forms.ModelForm):
+    class Meta:
+        model = DestinationConfig
+        fields = ["media", "label"]
+
+
 class NotificationMedium(ABC):
+    """
+    Must be defined by subclasses:
+
+    Class attributes:
+
+    - MEDIA_SLUG: short string id for the medium, lowercase
+    - MEDIA_NAME: human friendly id for the medium
+    - MEDIA_SETTINGS_KEY: the field in settings that is specific for this medium
+    - MEDIA_JSON_SCHEMA: A json-schema to describe the settings field to
+      javascript, used by the API
+    - Form: a django.forms.Form that validates the contents of the
+      settings-field. MEDIA_SETTINGS_KEY must be the field-name of a required
+      field in the form.
+
+    Class methods:
+
+    - send(event, destinations): How to send the given event to the given
+      destinations of type MEDIA_SLUG.
+    """
+
+    error_messages = {
+        "readonly_medium": "Media cannot be updated, only settings.",
+        "readonly_user": "User cannot be changed, only settings.",
+        "settings_type": "Settings has to be a dictionary.",
+    }
+
     class NotDeletableError(Exception):
         """
         Custom exception class that is raised when a destination cannot be
         deleted
         """
 
     @classmethod
-    @abstractmethod
-    def validate(cls, instance: RequestDestinationConfigSerializer, dict: dict, user: User) -> dict:
+    def validate(
+        cls, data: dict, user: User, instance: DestinationConfig = None, exception_class=ValidationError
+    ) -> dict:
+        if instance:
+            if data.get("media", "") != instance.media.slug:
```

Comment:
But in the case of updating an instance (PATCH) when no `media` is given then this will raise this error
```suggestion
            if "media" in data.keys() and data["media"] != instance.media.slug:
```

Reply:
It would be incorrect to fix this here as validate cannot know whether we are patching or not. The form is not and cannot be valid if media does not exist.

The immediate correct fix is in the API, the technically correct fix is also in he API, by making "media" part of the url. Both are out of scope.

Reply:
Alright, then you can create a follow up issue and we can merge this in

Reply:
I found a way, and discovered a new issue in the process: the API does not behave correctly when PATCH-ing the email-address unchanged, it only complains about 

```
{
  "non_field_errors": [
    "email_address"
  ]
}
```

.. stripping away the detail of the error for the email-address.

It would probably be better to instead return 200 Ok and pretend the PATCH never happened.

Reply:
.. and thinking some more, we don't need the medium in the url.